incubator/hnc/internal/reconcilers/object_test.go

```diff
@@ -15,6 +15,184 @@ import (
+			notWant:      []string{c1, c2},
+		}, {
+			name:         "not propagate object to neither negatively selected or treeSelected namespaces",
+			selector:     "!" + c1 + api.LabelTreeDepthSuffix,
+			treeSelector: "!" + c2,
+			want:         []string{c3},
+			notWant:      []string{c1, c2},
+		}, {
+			name:         "only propagate object to the intersection of selected and treeSelected namespaces",
+			selector:     c1 + api.LabelTreeDepthSuffix,
+			treeSelector: c2,
+			want:         []string{},
+			notWant:      []string{c1, c2, c3},
+		}}
+
+		for _, tc := range tests {
+			// Making a local copy of tc is necessary to ensure the correct value is passed to the closure,
+			// for more details look at https://onsi.github.io/ginkgo/ and search for 'closure'
+			tc := tc
+			It("Should handle exceptions: "+tc.name, func() {
+				// Set up namespaces
+				names := map[string]string{
+					p:  createNS(ctx, "parent"),
+					c1: createNS(ctx, "child1"),
+					c2: createNS(ctx, "child2"),
+					c3: createNS(ctx, "child3"),
+				}
+				setParent(ctx, names[c1], names[p])
+				setParent(ctx, names[c2], names[p])
+				setParent(ctx, names[c3], names[p])
+
+				tc.selector = replaceStrings(tc.selector, names)
+				tc.treeSelector = replaceStrings(tc.treeSelector, names)
+
+				// Create a Role with the selector and treeSelector annotation
+				makeObjectWithAnnotation(ctx, "Role", names[p], "testrole", map[string]string{
+					"propagate.hnc.x-k8s.io/select":     tc.selector,
+					"propagate.hnc.x-k8s.io/treeSelect": tc.treeSelector,
+				})
+				for _, ns := range tc.want {
+					ns = replaceStrings(ns, names)
+					Eventually(hasObject(ctx, "Role", ns, "testrole")).Should(BeTrue(), "When propagating testrole to %s", ns)
+				}
+				for _, ns := range tc.notWant {
+					ns = replaceStrings(ns, names)
+					Consistently(hasObject(ctx, "Role", ns, "testrole")).Should(BeFalse(), "When propagating testrole to %s", ns)
+				}
+			})
+		}
+	})
+
+	Context("Update exception annotations", func() {
+		const (
+			p  = "parent"
+			c1 = "child1"
+			c2 = "child2"
+			c3 = "child3"
+		)
+		tests := []struct {
+			name         string
+			selector     string
+			treeSelector string
+			want         []string
+			notWant      []string
+		}{{
+			name:         "update select",
+			selector:     "!" + c1 + api.LabelTreeDepthSuffix,
+			treeSelector: "",
+			want:         []string{c2, c3},
+			notWant:      []string{c1},
+		}, {
+			name:         "update treeSelect",
+			selector:     "",
+			treeSelector: "!" + c1,
+			want:         []string{c2, c3},
+			notWant:      []string{c1},
+		}}
+
+		for _, tc := range tests {
+			// Making a local copy of tc is necessary to ensure the correct value is passed to the closure,
+			// for more details look at https://onsi.github.io/ginkgo/ and search for 'closure'
+			tc := tc
+			It("Should handle exceptions: "+tc.name, func() {
+				// Set up namespaces
+				names := map[string]string{
+					p:  createNS(ctx, "parent"),
```

Comment:
Should be: `p: createNS(ctx, p)` since `p` already has the right value (also on line 90)

Reply:
done